diff --git a/patchlevel.h b/patchlevel.h
--- a/patchlevel.h
+++ b/patchlevel.h
@@ -25,6 +25,6 @@
    regexp `^#define[ 	]*PATCHLEVEL', since that's what support/mkversion.sh
    looks for to find the patch level (for the sccs version string). */
 
-#define PATCHLEVEL 31
+#define PATCHLEVEL 32
 
 #endif /* _PATCHLEVEL_H_ */
diff --git a/print_cmd.c b/print_cmd.c
--- a/print_cmd.c
+++ b/print_cmd.c
@@ -356,7 +356,9 @@ make_command_string_internal (command)
 	  break;
 
 	case cm_coproc:
-	  cprintf ("coproc %s ", command->value.Coproc->name);
+	  cprintf ("coproc ");
+	  if (command->value.Coproc->command->type != cm_simple)
+	    cprintf ("%s ", command->value.Coproc->name);
 	  skip_this_indent++;
 	  make_command_string_internal (command->value.Coproc->command);
 	  break;
